ttkwidgets/itemscanvas.py

```diff
@@ -103,7 +105,25 @@ def left_motion(self, event):
         rectangle = self.items[item]
         self.config(cursor="exchange")
         self.canvas.itemconfigure(item, fill="blue")
-        x, y = self.canvas.canvasx(event.x), self.canvas.canvasy(event.y)
+        xc, yc = self.canvas.canvasx(event.x), self.canvas.canvasy(event.y)
+        dx, dy = xc - self.current_coords[0], yc - self.current_coords[1]
+        self.current_coords = xc, yc
+#        x = self._max_x if x > self._max_x else x
+#        y = self._max_y if y > self._max_y else y
+#        x = 0 if x < 0 else x
+#        y = 0 if y < 0 else y
+        self.canvas.move(item, dx, dy)
+        self.canvas.move(rectangle, dx, dy)
+#        x1, y1, x2, y2 = self.canvas.coords(rectangle)
+#        if x1 < 0:
+#            x1, x2 = 0, x2 - x1
+#        if y1 < 0:
+#            y1, y2 = 0, y2 - y1
+#        if x2 > self._canvaswidth:
+#            x1, x2 = self._canvaswidth - (x2 - x1), self._canvaswidth
+#        if y2 > self._canvaswidth:
+#            y1, y2 = self._canvaswidth - (y2 - y1), self._canvaswidth
+        x, y = self.canvas.coords(item)
```

Comment:
Given how the item text and rectangle are moved instead of given a new position, I think the `xc` and `yc` values could be delimited beforehand to limit the functions called on the actual canvas to the two `.move` calls, so the position is not delimited after moving. Perhaps something like this?

```python
self.canvas.itemconfigure(item, fill="blue")
xc, yc = self.canvas.canvasx(event.x), self.canvas.canvasy(event.y)
xc, yc = max(min(xc, self._max_x), 0), max(min(yc, self._max_y), 0)
dx, dy = xc - self.current_coords[0], yc - self.current_coords[1]
self.current_coords = xc, yc
self.canvas.move(item, dx, dy)
self.canvas.move(rectangle, dx, dy)
```
Admittedly, it is only slightly more efficient. It might also be an option to put the delimiting `.coords` calls into an `if`-clause. What do you think?

Reply:
I have tested your code and I think the behavior is a bit strange because the dragging limits depend on where the user clicked on the item. Otherwise I see two ways of improving the function:
1) Remove `self.canvas.move(rectangle, dx, dy)` because it is set in place with `self.canvas.coords(rectangle, self.canvas.bbox(item))`

``` python
        self.canvas.itemconfigure(item, fill="blue")
        xc, yc = self.canvas.canvasx(event.x), self.canvas.canvasy(event.y)
        dx, dy = xc - self.current_coords[0], yc - self.current_coords[1]
        self.current_coords = xc, yc
        self.canvas.move(item, dx, dy)
        # check whether the new position of the item respects the boundaries
        x, y = self.canvas.coords(item)
        x, y = max(min(x, self._max_x), 0), max(min(y, self._max_y), 0)
        self.canvas.coords(item, x, y)
        self.canvas.coords(rectangle, self.canvas.bbox(item))
```
2) Use if statements to avoid replacing the item uselessly

``` python
        self.canvas.itemconfigure(item, fill="blue")
        xc, yc = self.canvas.canvasx(event.x), self.canvas.canvasy(event.y)
        dx, dy = xc - self.current_coords[0], yc - self.current_coords[1]
        self.current_coords = xc, yc
        self.canvas.move(item, dx, dy)
        self.canvas.move(rectangle, dx, dy)
        # check whether the new position of the item respects the boundaries
        x, y = self.canvas.coords(item)
        x2, y2 = max(min(x, self._max_x), 0), max(min(y, self._max_y), 0)
        if x != x2 or y != y2:
            self.canvas.coords(item, x2, y2)
            self.canvas.coords(rectangle, self.canvas.bbox(item))
```
Tell me which one you prefer or propose something else.

Reply:
The first seems the most readable, as (at least to me), it's immediately clear what's happening. It's a small optimization, but I think it's worth it. You're right about the boundaries in my code, that is inconsistent and thus not good.

Reply:
I have committed the changes.